### PR TITLE
[onnx] Lowering `onnx.dequantize_linear` to `torch`

### PR DESCRIPTION
**Pull request by rsuderman**
We can make the per-tensor version of the operation to the dequantize operation via marking with the make quantized tensor component. This introductions the `qint*` and `quint*` tensor type that can be lowered to teh appropriate dequantization behavior during the torch-to-linalg conversion.

### CODE REVIEW

**Comment by vivekkhandelwal1** [comment]
Hi @rsuderman, I had taken this op for the lowering. I'm not sure if you went through this https://github.com/nod-ai/SHARK-Turbine/issues/334 or not.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp [hunk 1]
```diff
@@ -1156,6 +1156,59 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
...
+          auto qTensorTy = rewriter.getType<Torch::ValueTensorType>(
+              resultType.getOptionalSizes(), qTy);
+          scale = rewriter.create<Torch::AtenItemOp>(
+              binder.getLoc(), rewriter.getType<Torch::FloatType>(), scale);
+          zeropoint = rewriter.create<Torch::AtenItemOp>(
+              binder.getLoc(), rewriter.getType<Torch::IntType>(), zeropoint);
+
+          auto quantize =
+              rewriter.create<Torch::Aten_MakePerTensorQuantizedTensorOp>(
+                  binder.getLoc(), qTensorTy, operand, scale, zeropoint);
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:1204]: I think this won't handle per-axis quantization tensor op.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp [hunk 2]
```diff
@@ -1156,6 +1156,59 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
...
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        Torch::ValueTensorType resultType;
+        llvm::SmallVector<Value> operands;
+        if (binder.tensorOperands(operands, 3) ||
+            binder.tensorResultType(resultType))
+          return failure();
+
+        Value operand = operands[0];
+        Value scale = operands[1];
+        Value zeropoint = operands[2];
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:1170]: Changed to pull the value conditionally.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp [hunk 3]
```diff
@@ -1156,6 +1156,59 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
...
+        auto operandTy = operand.getType().cast<Torch::ValueTensorType>();
+
+        auto scaleTy = scale.getType().dyn_cast<Torch::ValueTensorType>();
+        if (!scaleTy || !scaleTy.hasSizes())
+          return rewriter.notifyMatchFailure(binder.op, "requires known rank");
+        if (!resultType.hasDtype())
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "requires known resulty dtype");
+
+        if (scaleTy.getSizes().size() == 0) {
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:1181]: It is not yet supported. In a follow up we can do a similar trick with `AtenMakePerChannelQuantzedOp` but per-tensor should be good enough for now.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp [hunk 4]
```diff
@@ -1156,6 +1156,59 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
...
+            return rewriter.notifyMatchFailure(binder.op,
+                                               "unsupported result dtype");
+          }
+
+          auto qTensorTy = rewriter.getType<Torch::ValueTensorType>(
+              resultType.getOptionalSizes(), qTy);
+          scale = rewriter.create<Torch::AtenItemOp>(
+              binder.getLoc(), rewriter.getType<Torch::FloatType>(), scale);
+          zeropoint = rewriter.create<Torch::AtenItemOp>(
+              binder.getLoc(), rewriter.getType<Torch::IntType>(), zeropoint);
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:1200]: Correct, however we verified above that the tensor is already rank-0.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp (hunk 1 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:1207]: Okay

**Review by vivekkhandelwal1 — APPROVED** [review]
LGTM

**Comment by rsuderman** [comment]
> Hi @rsuderman, I had taken this op for the lowering. I'm not sure if you went through this [nod-ai/SHARK-Turbine#334](https://github.com/nod-ai/SHARK-Turbine/issues/334) or not.

Generally the quantization work needs to be separated out from the general op-by-op lowering work. Because `torch` depends on sidechanneling quantization information it is likely we will need passes to propagate / setup the sidechannelled zero-point / scale information.

**Pull request merged**